### PR TITLE
Feat scale based on container count

### LINKED ISSUE #5546: Allow Addon Resizer to scale based on container count
How much load some workload see depends on number of containers, not on the number of nodes. Number of nodes is not good enough predictor of how many containers are in the cluster and so we should allow scaling directly based on the number of containers.

I want to implement this on 1.8 branch.

### PR DESCRIPTION
**Pull request by AhmedGrati**
#### What type of PR is this?

/kind feature
#### What this PR does / why we need it:
Mainly related to #5547. Adds scaling based on container count.

#### Which issue(s) this PR fixes:

Fixes #5546.

#### Special notes for your reviewer:
It's my first contribution to the project, so please bear with me. Also this is a docker image with this feature:
`ahmedgrati/addon-resizer-amd64:1.8.19`. You can use it for manual testing.

#### Does this PR introduce a user-facing change?

```release-note
Add scaling mode option with the flag --scaling-mode which accepts two options: 'node-proportional' or 'container-proportional'.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

- KEP: https://github.com/kubernetes/autoscaler/blob/master/addon-resizer/enhancements/5546-scaling-based-on-container-count/README.md

### CODE REVIEW

File: addon-resizer/dep.yaml
Comment (jbartosik) [inline addon-resizer/dep.yaml:1]: Please don't add this file only to remove it in a later commit

File: addon-resizer/nanny/kubernetes_client.go [hunk 1]
```diff
@@ -57,11 +60,35 @@ type kubernetesClient struct {
...
+	return k.countContainersThroughAPI()
+}
+
+func (k *kubernetesClient) countContainersThroughAPI() (uint64, error) {
+	// Set ResourceVersion = 0 to use cached versions.
+	options := metav1.ListOptions{
+		ResourceVersion: "0",
+		FieldSelector:   getPodSelectorExcludingDonePodsOrDie(),
+	}
+	pods, err := k.clientset.CoreV1().Pods("").List(context.TODO(), options)
```
Comment (jbartosik) [inline addon-resizer/nanny/kubernetes_client.go:84]: Please use [NamespaceAll](https://github.com/kubernetes/autoscaler/blob/4cb926b2f0bb4e4b5fc206bf5680821311b09487/addon-resizer/vendor/k8s.io/api/core/v1/types.go#L30C2-L30C14) const to make this easier to understand

```suggestion
	pods, err := k.clientset.CoreV1().Pods(corev1.NamespaceAll).List(context.TODO(), options)
```

File: addon-resizer/nanny/kubernetes_client.go [hunk 2]
```diff
@@ -103,13 +130,14 @@ func extractMetricValueForNodeCount(mf dto.MetricFamily, metricName string) (uin
 		if *metric.Gauge.Value < 0 {
 			return 0, fmt.Errorf("%s: metric unknown", metricName)
 		}
+		fmt.Println(metric)
```
Comment (jbartosik) [inline addon-resizer/nanny/kubernetes_client.go:133]:
```suggestion
```

Looks like debugging leftover, please remove

File: addon-resizer/nanny/main/pod_nanny.go [hunk 3]
```diff
@@ -60,9 +64,10 @@ var (
...
 	estimator      = flag.String("estimator", "linear", "The estimator to use. Currently supported: linear, exponential")
-	minClusterSize = flag.Uint64("minClusterSize", 16, "The smallest number of nodes resources will be scaled to. Must be > 1. This flag is used only when an exponential estimator is used.")
-	useMetrics     = flag.Bool("use-metrics", false, "Whether to use apiserver metrics to detect cluster size instead of the default method of listing node objects from the Kubernetes API.")
+	minClusterSize = flag.Uint64("minClusterSize", 16, "The smallest number of resources will be scaled to. Must be > 1. This flag is used only when an exponential estimator is used.")
+	useMetrics     = flag.Bool("use-metrics", false, "Whether to use apiserver metrics to detect cluster size instead of the default method of listing objects from the Kubernetes API.")
 	hcAddress      = flag.String("healthcheck-address", ":8080", "The address to expose an HTTP health-check on.")
+	scalingMode    = flag.String("scaling-mode", "node-proportional", "The mode of scaling to be used. Possible values: 'node-proportional' or 'container-proportional'")
```
Comment (jbartosik) [inline addon-resizer/nanny/main/pod_nanny.go:70]:
```suggestion
	scalingMode    = flag.String("scaling-mode", NodeProportional, "The mode of scaling to be used. Possible values: 'node-proportional' or 'container-proportional'")
```

File: addon-resizer/nanny/nanny_lib.go [hunk 4]
```diff
@@ -33,9 +33,11 @@ import (
...
 
 const (
-	unknown   operation = iota
-	scaleDown operation = iota
-	scaleUp   operation = iota
+	unknown               operation = iota
+	scaleDown             operation = iota
+	scaleUp               operation = iota
+	ContainerProportional           = "container-proportional"
+	NodeProportional                = "node-proportional"
```
Comment (jbartosik) [inline addon-resizer/nanny/nanny_lib.go:40]: Please define those consts only in one place (here or in pod_nanny.go)

File: addon-resizer/nanny/nanny_lib_test.go [hunk 5]
```diff
@@ -232,34 +232,35 @@ func TestUpdateResources(t *testing.T) {
...
+		scalingMode string
+		want        updateResult
 	}{
 		// No changes to the resources
-		{0, standard, standard, now, noDelay, noDelay, noChange},
-		{0, standard, standard, oneHourAgo, noDelay, noDelay, noChange},
-		{0, standard, standard, oneHourAgo, oneMinuteDelay, noDelay, noChange},
-		{0, standard, standard, oneHourAgo, noDelay, oneMinuteDelay, noChange},
-		{10, standard, siStandard, now, noDelay, noDelay, noChange},
+		{0, standard, standard, now, noDelay, noDelay, NodeProportional, noChange},
```
Comment (jbartosik) [inline addon-resizer/nanny/nanny_lib_test.go:244]: Please keep existing tests for node-proportional and add new ones for container-proportional

**Review by jbartosik — COMMENTED** [review]
Thank you, mostly looks good I left some comments

Did you tests the change? If so please let me know what tests did you perform

**Comment by jbartosik** [comment]
@laoj2 @kwiesmueller FYI

File: addon-resizer/Makefile
[the remark this answers is not in the document]
Comment (AhmedGrati) [reply addon-resizer/Makefile:1]: Don't know how, but I've deleted these changes before opening the PR.

**Comment by AhmedGrati** [comment]
Thanks for the review @jbartosik. I updated the PR.
The test that I tried is the following:
```
- Starting the deployment scaled by Addon Resizer, based on node count.
- Observe the size of the deployment and that it's stable.
- Start deployment with 100 pause containers.
- Observe the scaled deployment change resources appropriately.
```
I used 10 pause containers instead of 100 since I don't have the resources to handle 100 containers.

**Comment by AhmedGrati** [comment]
ping @jbartosik. PTAL

File: addon-resizer/Makefile
[the remark this answers is not in the document]
Comment (AhmedGrati) [reply addon-resizer/Makefile:1]: Addressed.

File: addon-resizer/dep.yaml
Comment (AhmedGrati) [reply addon-resizer/dep.yaml:1]: Addressed.

File: addon-resizer/nanny/kubernetes_client.go (hunk 1 above)
Comment (AhmedGrati) [reply addon-resizer/nanny/kubernetes_client.go]: Addressed.

File: addon-resizer/nanny/kubernetes_client.go (hunk 2 above)
Comment (AhmedGrati) [reply addon-resizer/nanny/kubernetes_client.go]: Addressed.

File: addon-resizer/nanny/main/pod_nanny.go (hunk 3 above)
Comment (AhmedGrati) [reply addon-resizer/nanny/main/pod_nanny.go]: Addressed.

File: addon-resizer/nanny/nanny_lib_test.go (hunk 5 above)
Comment (AhmedGrati) [reply addon-resizer/nanny/nanny_lib_test.go:244]: Addressed.

**Comment by AhmedGrati** [comment]
@jbartosik all of the comments are addressed. PTAL when you have time for it.

**Comment by AhmedGrati** [comment]
PING @jbartosik.

**Review by jbartosik — APPROVED** [review]
/lgtm
/approve

**Pull request merged**